src/System.Private.ServiceModel/src/Internals/System/Runtime/TimeoutHelper.cs

```diff
@@ -1,12 +1,69 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System.Collections.Concurrent;
 using System.Diagnostics.Contracts;
 using System;
 using System.Threading;
+using System.Threading.Tasks;
 
 namespace System.Runtime
 {
+    class CoalescedTimeoutTokens
+
+    {
+        private readonly uint _coalescingSpanMs;
+        private readonly ConcurrentDictionary<long, Task<CancellationToken>> _timeToToken =
+            new ConcurrentDictionary<long, Task<CancellationToken>>();
+
+        public CoalescedTimeoutTokens(int coalescingSpanMs = 15)
+        {
+            _coalescingSpanMs = (uint)coalescingSpanMs;
+        }
+
+        public CancellationToken FromTimeout(int millisecondsTimeout)
+        {
+            return FromTimeoutAsync(millisecondsTimeout).Result;
+        }
+
+        public Task<CancellationToken> FromTimeoutAsync(int millisecondsTimeout)
```

Comment:
Why have tasks involved?  It looks like the only consumer of this always immediately gets the Result to extract the CancellationToken.  It'd be cheaper to not allocate new tasks on each call if those tasks aren't necessary.